### ISSUE TITLE
Base href URL : How to add a prefix to all URLs ?

### ISSUE THREAD

**Issue by samfrach** [preboot/angular2-webpack#222]
Hello,

I would like to serve my application not from the server root but from /myapp/

I've added a <base href="/myapp/"> in the index.html file but javascript files are still generated in the server root...

where do I configure that ?

Furthermore, where can I configure the socket that refreshes the page ? I'd like to change the IP and port (I want to configure a nginx server on a machine that proxies to node on a virtual machine ... yes for dev as well)

thank you

**Comment by mstrzoda** [comment]
i have the same question, any ideas?

**Comment by ollwenjones** [comment]
> I would like to serve my application not from the server root but from `/myapp/`
I don't have time to test changing this, but I believe this is what the `<base href="/">` in the index.html is for. You'll likely have to update some other configuration to match.

> Furthermore, where can I configure the socket that refreshes the page ? 
The webpack-dev-server configuration is in the wepback config in a couple of places:

```JavaScript
  /**
   * Output. Add chunk hashes for production.
   * Reference: http://webpack.github.io/docs/configuration.html#output
   */
  config.output = isTest ? {} : {
    path: root('target/dist'),
    publicPath: isProd ? '/' : 'http://localhost:8080/',
    filename: isProd ? 'js/[name].[hash].js' : 'js/[name].js',
    // source maps for production?
    // sourceMapFilename: isProd ? '[name].map' : '[name].[hash].map',
    chunkFilename: isProd ? '[id].[hash].chunk.js' : '[id].chunk.js'
  };
```

```JavaScript
  /**
   * Dev server configuration
   * Reference: http://webpack.github.io/docs/configuration.html#devserver
   * Reference: http://webpack.github.io/docs/webpack-dev-server.html
   */
  config.devServer = {
    contentBase: './src/public',
    historyApiFallback: true,
    stats: 'minimal', // none (or false), errors-only, minimal, normal (or true) and verbose
    host: process.env.host || 'localhost',
    port: process.env.port || 8080,
    proxy: getProxy()
  };
```

There's pretty extensive reference linking in that file. You should check it out. (Don't treat it as a black box!)

**Comment by herb-stone** [comment]
I have the same issue. Deploying my project directly to a tomcat webapps directory works fine. But I need to host several webapps each in its own subdirectory. Appending the path 
`path: root('my-app/dist'),`
 dosn't  work for me. 
Any suggestions?

**Comment by mcescalante** [comment]
I changed the base URL in 2 places to get it to work successfully for me:

1. Line 62 in `webpack.config.js`. Change the publicPath `'/'` to your base:
```js
  config.output = isTest ? {} : {
    path: root('dist'),
    publicPath: isProd ? '/' : 'http://localhost:8080/', //<-- change the '/' in this line to '/yourapp'
    filename: isProd ? 'js/[name].[hash].js' : 'js/[name].js',
    chunkFilename: isProd ? '[id].[hash].chunk.js' : '[id].chunk.js'
  };
```

2. In `src/public/index.html` on line 8 change the `<base href>` tag:
```html
<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <title>Angular 2 App | ng2-webpack</title>
    <link rel="icon" type="image/x-icon" href="/img/favicon.ico">

    <base href="/"> <!-- change this / to /yourapp -->
  </head>
  <body>
    <my-app>Loading...</my-app>
  </body>
</html>
```

Restart webpack and this change should reflect in both your `npm start` and `npm build` commands

**Comment by babelouest** [comment]
I have the same problem, @mcescalante 's solution works fine except for `public` folder, all the css and js files return an error 404.
For example, a css file url is `http://localhost:8080/yourapp/css/bootstrap.css`, but it's still accessible to `http://localhost:8080/css/bootstrap.css`

**Comment by ollwenjones** [comment]
@babelouest We recently ran into this, and I had forgotten all about it. I'm pretty sure what's missing is some `to` setting in the [CopyWebpackPlugin](https://github.com/kevlened/copy-webpack-plugin) - since that is what is putting those static assets in the right place.

**Comment by ollwenjones** [comment]
My mistake, it looks like CopyWebpackPlugin is only used in production, which means it isn't the problem in development.

**Comment by mcescalante** [comment]
@babelouest good catch, I missed that for some reason. Just tried this out and I think it fixes what you reported there:

On line 62 change both the `/` and the `http://localhost:8080/` and add the new base URL onto both of them. That adjusted the static file location for me so that it was`/yourapp/css/test.css` and **not** `/css/test.css. 

Example:

```
  .output = isTest ? {} : {
    path: root('dist'),
    publicPath: isProd ? '/yourapp' : 'http://localhost:8080/yourapp',
    filename: isProd ? 'js/[name].[hash].js' : 'js/[name].js',
    chunkFilename: isProd ? '[id].[hash].chunk.js' : '[id].chunk.js'
  };
```

Let me know if this is the desired behavior and I'll update my answer above as well.

**Comment by babelouest** [comment]
Thanks @mcescalante , but it's the config I had and it's still not working to access `public` folder files.

In fact, since I may access the dev instance from other devices, my publicPath is just:
```
publicPath: '/',
```

and if I change it with
```
publicPath: '/yourapp',
```
and update the `base href` tag in the index.html file, I have the error 404 with the public files as described below.

**Comment by mcescalante** [comment]
@babelouest that's bizarre, can you put your changed `webpack.config.js` and `index.html` into a gist so I can take a look to help?

**Comment by babelouest** [comment]
Better than a gist, here is the source of my app.

`webpack.config.js`: https://github.com/babelouest/hutch/blob/master/webapp/webpack.config.js
`index.html`: https://github.com/babelouest/hutch/blob/master/webapp/src/public/index.html

**Comment by herb-stone** [comment]
@babelouest, prefix the pathes in your index.html with slashes: href="/css ...".

**Comment by babelouest** [comment]
@herb-stone , it still doesn't solve the case, unfortunately.

**Comment by sadidkhan** [comment]
@babelouest how did you solve it?

**Comment by babelouest** [comment]
@sadidkhan , I didn't solve it, the generated web site is still on an absolute path.

**Comment by herb-stein** [comment]
@sadidkhan use the angular cli compiler option ``--baseHref=baseHref`` to set the base url for the application being built.

see: https://angular.io/cli/build
